allennlp/data/semparse/type_declarations/type_declaration.py

```diff
@@ -254,3 +257,128 @@ def make_VariableExpression(self, name):
             else:
                 raise RuntimeError("Unknown prefix: %s. Did you forget to pass it to the constructor?" % prefix)
         return super(DynamicTypeLogicParser, self).make_VariableExpression(name)
+
+
+def _substitute_any_type(_type: Type, basic_types: Set[BasicType]) -> Set[Type]:
+    """
+    Takes a type, and a set of basic types and substitutes all instances of ANY_TYPE with all possible basic
+    types, and returns a set with all possible combinations.
+    Note that this substitution is unconstrained. That is, If you have a type with placeholders,
+    <#1,#1> for example, this may substitute the placeholders with different basic types. In that case, you'd
+    want to use ``_substitute_placeholder_type`` instead.
+    """
+    if _type == ANY_TYPE:
+        return basic_types
+    if isinstance(_type, (BasicType, PlaceholderType)):
+        return set([_type])
+    substitutions = set()
+    for first_type in _substitute_any_type(_type.first, basic_types):
+        for second_type in _substitute_any_type(_type.second, basic_types):
+            substitutions.add(ComplexType(first_type, second_type))
+    return substitutions
+
+
+def _substitute_placeholder_type(_type: Type, basic_type: BasicType) -> Type:
+    """
+    Takes a type with placeholders, and a basic type and substitutes all occurrences of the placeholder with
+    that type.
+    """
+    # TODO (pradeep): This assumes there's just one placeholder in the type. So this doesn't work with
```

Comment:
Is there a way to make this crash if there is more than one placeholder?

Reply:
Done.